src/index.ts

```diff
@@ -1,65 +1,238 @@
+const ALLOW_REQUEST_HEADERS = [
+  // NOTE: unclear if routing layer will handle these for us, so for now update this list as needed
+  'Content-Type',
+  'Authorization',
+  'X-Api-Key',
+  'X-Hiro-Product',
+  'X-Hiro-Version',
+];
 
-  await server.register(cors);
+// Access-Control-Allow-Headers
+const ALLOW_RESPONSE_HEADERS = [
+  // Allow browser JS to read these response headers
+  'Content-Type',
+  'X-Api-Version',
+];
 
-  /*
-  server.addContentTypeParser(
-    'application/octet-stream',
-    (req, payload, done) => {
-      done(null, payload);
-    }
+async function postTxMulticast(
+  req: FastifyRequest,
+  reply: FastifyReply,
+  res: Readable
+) {
+  const contentType = req.headers['content-type'];
+  await performTxMulticast(req.body as Buffer, contentType);
+}
+
+/**
+ * Reads the upstream response body to memory and logs it.
+ * @param res Response from upstream
+ * @returns a new readable stream that can be used to send the response body downstream.
+ */
+function addResponseLogging(
+  req: FastifyRequest,
+  reply: FastifyReply,
+  res: Readable
+) {
+  return readPassthrough(res, (body) => {
+    const bodyString = body.toString('utf8');
+    logger.debug(
+      {
+        method: req.method,
+        url: req.url,
+        status: reply.statusCode,
+        response: bodyString,
+      },
+      `Response to ${req.method} ${req.url} returned ${reply.statusCode}: ${bodyString}`
+    );
+  });
+}
+
+function addRequestLogging(req: FastifyRequest) {
+  const bodyString =
+    (req.body as Buffer)?.toString('utf8') ?? '<no request body>';
+  logger.debug(
+    {
+      method: req.method,
+      url: req.url,
+      request: bodyString,
+    },
+    `Request to ${req.method} ${req.url}: ${bodyString}`
   );
+  return req;
+}
 
-  server.addContentTypeParser('application/json', (req, payload, done) => {
-    done(null, payload);
+function readPassthrough(readable: Readable, onBody: (body: Buffer) => void) {
+  const chunks: Buffer[] = [];
+  const transform = new Transform({
+    autoDestroy: true,
+    emitClose: true,
+    transform(chunk, _encoding, callback) {
+      chunks.push(chunk);
+      callback(null, chunk);
+    },
+  });
+  transform.on('close', () => {
+    const body = chunks.length === 1 ? chunks[0] : Buffer.concat(chunks);
+    onBody(body);
   });
+  return readable.pipe(transform);
+}
+
+function handleProxyResponse(
+  request: FastifyRequest,
+  reply: FastifyReply,
+  response: Readable
+) {
+  // Handle request/response logging
+  if (ENV.LOG_REQUESTS) {
+    request = addRequestLogging(request);
+  }
+  if (ENV.LOG_RESPONSES) {
+    response = addResponseLogging(request, reply, response);
+  }
+
+  // Handle tx-multicast config
+  if (
+    isTxMulticastEnabled() &&
+    request.method === 'POST' &&
+    request.url === '/v2/transactions'
+  ) {
+    postTxMulticast(request, reply, response).catch((error) => {
+      logger.error(error, `Error performing tx-multicast: ${error}`);
+    });
+  }
```

Comment:
Nit: these could be registered as separate Fastify plugins when you build the server if these conditions are met... No real functionality change, just a cleaner use of Fastify

Reply:
The issue I ran into initially with trying that (a separate handler for POST /v2/transactions at least) was that I have to re-implement some of the `@fastify/http-proxy` internal logic for the handler (seems to use fastify-reply-from under the hood but with a few other things). I couldn't get it working correctly -- but good area for potential improvement later.